File: SemanticNetwork/src/main/java/de/kimanufaktur/nsm/decomposition/graph/spreadingActivation/MarkerPassing/DoubleMarkerPassing.java
/*
 * Copyright (C) Johannes Fähndrich - All Rights Reserved.
 * Unauthorized copying of this file, via any medium is strictly
 * prohibited Proprietary and confidential.
 * Written by Johannes Fähndrich <faehndrich@gmail.com.com>,  2011
 */

package de.kimanufaktur.nsm.decomposition.graph.spreadingActivation.MarkerPassing;

import com.google.common.collect.BiMap;
import com.google.common.collect.HashBiMap;
import de.kimanufaktur.nsm.decomposition.Concept;
import de.kimanufaktur.nsm.decomposition.Decomposition;
import de.kimanufaktur.nsm.decomposition.Definition;
import de.kimanufaktur.nsm.graph.entities.links.*;
import de.kimanufaktur.nsm.graph.entities.nodes.DoubleNodeWithMultipleThresholds;
import de.kimanufaktur.markerpassing.*;
import org.jgrapht.Graph;

import java.lang.reflect.InvocationTargetException;
import java.util.*;

/**
 * Created by faehndrich on 06.05.15.
 */
public class DoubleMarkerPassing extends SpreadingAlgorithm {



    int pulsecount = 0;

    double currentDoubleActivation = 0.0;
    double maximalDoubleActivation = 0.0;

    Map<Concept, Double> doubleActivation = new HashMap<>();
    BiMap<Concept, Node> nodes = HashBiMap.create();
    List<Marker> originMarkerClasses = new ArrayList<>();

    public <T extends DoubleNodeWithMultipleThresholds> DoubleMarkerPassing(Graph graph, Map<Concept, Double> threshold, Class<T> nodeType) {
        //set active nodes
        final HashSet<Node> activenodes = new HashSet<Node>();
        this.setActiveNodes(activenodes);
        //set firing nodes
        HashSet<Node> firingnodes = new HashSet<>();
        this.setFiringNodes(firingnodes);
        //fill node network with nodes;
//        for (Map.Entry<Concept,Double> entry : threshold.entrySet()){
//            System.out.println("DEBUG lökojö: "+entry.getKey().getLitheral()+" "+entry.getValue());
//        }
        fillNodes(graph, threshold, nodeType);

        //Create termination condition
        TerminationCondition terminationCondition = new TerminationCondition() {
            @Override
            public boolean compute() {
                double lastDoubleActivation = currentDoubleActivation;

                currentDoubleActivation = 0.0;

                for (Node activeNode : getActiveNodes()) {
                    if (activeNode instanceof DoubleNodeWithMultipleThresholds) {
                        currentDoubleActivation += ((DoubleNodeWithMultipleThresholds) activeNode).getDoubleActivation();
                    }
                }
                if (currentDoubleActivation > maximalDoubleActivation) {
                    maximalDoubleActivation = currentDoubleActivation;
                }
                if (currentDoubleActivation >= MarkerPassingConfig.getDoubleActivationLimit()) {
                    return true;
                } /*else if (lastDoubleActivation != 0.0 && lastDoubleActivation == currentDoubleActivation) {
                    return true;
                }*/ else {
                    return ++pulsecount >= MarkerPassingConfig.getTerminationPulsCount();
                }


            }
        };
        this.setTerminationCondition(terminationCondition);

        //Create a SelectFiringNodesFunction which add all nodes which are over the threshold
        SelectFiringNodesFunction selectFiringNodesFunction = new SelectFiringNodesFunction() {
            @Override
            public Collection<Node> compute(Collection<Node> list) {
                List<Node> firingNodes = new ArrayList<>();
                for (Node node : list) {
                    if (node != null) {
                        if (node.checkThresholds(originMarkerClasses)) {
                            firingNodes.add(node);
                        }
                    }
                }
                return firingNodes;
            }
        };
        this.setSelectFiringNodes(selectFiringNodesFunction);

        //Create in-function
        InFunction inFunction = new InFunction() {
            @Override
            public void compute(Collection<SpreadingStep> list, Node node) {
                if (node instanceof DoubleNodeWithMultipleThresholds) {
                    ((DoubleNodeWithMultipleThresholds) node).in(list);
                    getActiveNodes().add(node);
                }
                //Node has gotton input, so it is a active node. The decision if the node will fire in the
                //next pulse, depends on the selection in the SelectFiringNodes function.
            }
        };
        this.setIn(inFunction);

        //Create out-function
        OutFunction outFunction = new OutFunction() {
            @Override
            public List<SpreadingStep> compute(Node node) {
                List<SpreadingStep> activationOutput = new ArrayList<>();
                if (node instanceof DoubleNodeWithMultipleThresholds) {
                    activationOutput.addAll(((DoubleNodeWithMultipleThresholds) node).out());
                }
                return activationOutput;
            }
        };
        this.setOut(outFunction);

        //Create pre-processing steps
        List<ProcessingStep> preProcessing = new ArrayList<>();
        this.setPreprocessingSteps(preProcessing);

        //Create post-processing steps
        List<ProcessingStep> postProcessing = new ArrayList<>();
        this.setPostprocessingSteps(postProcessing);
    }

    public static void doInitialMarking(List<Map<Concept, List<? extends Marker>>> startActivation, DoubleMarkerPassing doubleMarkerPassingAlgo) {
        for (Map<Concept, List<? extends Marker>> m : startActivation) {
            for (Map.Entry<Concept, List<? extends Marker>> e : m.entrySet()) {

                // 4 testing
//            	System.out.print("doInitialMarking startActivation concept: " + e.getKey().getLitheral());

                for (Marker marker : e.getValue()) {
                    doubleMarkerPassingAlgo.addMarkerToNode(e.getKey(), marker);

                    // 4 testing
//                    DoubleMarkerWithOrigin tmp = (DoubleMarkerWithOrigin)marker;
//                    System.out.println("; set marker activation: " + tmp.getActivation());

                    //add start markers to threshold
                    doubleMarkerPassingAlgo.getOriginMarkerClasses().add(marker);
                }
            }
        }
        // 4 testing
//        List<Marker> tmpmarker = doubleMarkerPassingAlgo.getOriginMarkerClasses();
//        for(Marker asdf : tmpmarker)
//        	System.out.println("doInitialMarking getOriginMarkerClasses are: " + ((DoubleMarkerWithOrigin)asdf).getOrigin().getLitheral());

    }

    public double getCurrentDoubleActivation() {
        return currentDoubleActivation;
    }

    public List<Marker> getOriginMarkerClasses() {
        return this.originMarkerClasses;
    }

    public double getMaximalDoubleActivation() {
        return maximalDoubleActivation;
    }

    public void setMaximalDoubleActivation(double maximalDoubleActivation) {
        this.maximalDoubleActivation = maximalDoubleActivation;
    }

    public Map<Concept, Double> getDoubleActivation() {
        return doubleActivation;
    }

    public void setDoubleActivation(Map<Concept, Double> doubleActivation) {
        this.doubleActivation = doubleActivation;
    }

    public BiMap<Concept, Node> getNodes() {
        return nodes;
    }

    public int getPulsecount() {
        return pulsecount;
    }

    public void setPulsecount(int pulsecount) {
        this.pulsecount = pulsecount;
    }

    public Node getNodeForConcept(Concept concept) {
        return nodes.get(concept);
    }

    public Concept getConceptForNode(Node node) {
        for (Map.Entry entry : nodes.entrySet()) {
            if (entry.getValue().equals(node)) {
                return (Concept) entry.getKey();
            }
        }
        return null;
    }

    /**
     * Fill the nodes for the activation spreading with the given graph. This graph is used as basis, for the
     * spreading activation. The graph is not altered, but its vertices are used to implement nodes for the spreading.
     *
     * @param graph     the graph to take the vertices and edges from. The vertices are concepts and the relations are links.
     * @param threshold the threshold to set for each node. TODO: have a sprecific threshold for each node type
     */
    public <T extends DoubleNodeWithMultipleThresholds> void fillNodes(Graph graph, Map<Concept, Double> threshold, Class<T> nodeType) {

        for (Concept concept : (Set<Concept>) graph.vertexSet()) {
            if (Decomposition.getConcepts2Ignore().contains(concept)) {
                continue;
            } else {
                Node node = nodes.get(concept);
                if (node == null || nodes.inverse().get(node).getDecompositionElementCount() < concept.getDecompositionElementCount()) {
                    Node tmpnode = addConceptRecursivly(concept, threshold, nodeType);
                    nodes.put(concept, tmpnode);
                }
            }
        }
        return;
    }

    /**
     * add the concept to the nodes network recursively
     *
     * @param concept the concept to add.
     */
    private <T extends DoubleNodeWithMultipleThresholds> Node addConceptRecursivly(Concept concept, Map<Concept, Double> threshold, Class<T> nodeType) {
        if (Decomposition.getConcepts2Ignore().contains(concept)) {
            return null;
        } else {
            T node = (T) nodes.get(concept);
            if (node == null) {
                try {
                    node = nodeType.getDeclaredConstructor(Concept.class).newInstance(concept);//concept.nodeType.newInstance(); //new DoubleNodeWithMultipleThresholds(concept);

                } catch (InstantiationException e) {
                    e.printStackTrace();
                } catch (IllegalAccessException e) {
                    e.printStackTrace();
                } catch (NoSuchMethodException e) {
                    e.printStackTrace();
                } catch (InvocationTargetException e) {
                    e.printStackTrace();
                }
                node.setThreshold(threshold);
                nodes.put(concept, node);
            } else if (nodes.inverse().get(node).getDecompositionElementCount() > concept.getDecompositionElementCount()) {
                return node;
            }
            for (Concept syn : concept.getSynonyms()) {
                if (syn != null && !Decomposition.getConcepts2Ignore().contains(concept)) {
                    Node synNode = nodes.get(syn);
                    if (synNode == null) {
                        synNode = addConceptRecursivly(syn, threshold, nodeType); //TODO: have a specific threshold for synonyms
                    }
                    if (synNode != null) {
                        SynonymLink link = new SynonymLink();
                        link.setSource(node);
                        link.setTarget(synNode);
                        link.setWeight(MarkerPassingConfig.getSynonymLinkWeight());
                        if (!node.getLinks().contains(link)) {
                            node.addLink(link);
                        }
                        SynonymLink backLink = new SynonymLink();
                        backLink.setSource(synNode);
                        backLink.setSource(node);
                        backLink.setWeight(MarkerPassingConfig.getSynonymLinkWeight());
                        if(!synNode.getLinks().contains(backLink)){
                            synNode.addLink(backLink);
                        }
                    }
                }
            }

            for (Concept hypo : concept.getHyponyms()) {
                if (hypo != null && !Decomposition.getConcepts2Ignore().contains(concept)) {
                    Node hypoNode = nodes.get(hypo);
                    if (hypoNode == null) {
                        hypoNode = addConceptRecursivly(hypo, threshold, nodeType); //TODO: have a specific threshold for hyponym
                    }
                    if (hypoNode != null) {
                        HyponymLink link = new HyponymLink();
                        link.setSource(node);
                        link.setTarget(hypoNode);
                        link.setWeight(MarkerPassingConfig.getHyponymLinkWeight());
                        if (!node.getLinks().contains(link)) {
                            node.addLink(link);
                        }
                    }
                }
            }
            for (Concept hyper : concept.getHypernyms()) {
                if (hyper != null && !Decomposition.getConcepts2Ignore().contains(concept)) {
                    Node hyperNode = nodes.get(hyper);
                    if (hyperNode == null) {
                        hyperNode = addConceptRecursivly(hyper, threshold, nodeType); //TODO: have a specific threshold for hypernyms
                    }
                    if (hyperNode != null) {
                        HypernymLink link = new HypernymLink();
                        link.setSource(node);
                        link.setTarget(hyperNode);
                        link.setWeight(MarkerPassingConfig.getHypernymLinkWeight());
                        if (!node.getLinks().contains(link)) {
                            node.addLink(link);
                        }
                    }
                }
            }
            for (Definition definition : concept.getDefinitions()) {
                for (Concept def : definition.getDefinition()) {
                    if (def != null && !Decomposition.getConcepts2Ignore().contains(concept)) {
                        Node defnode = nodes.get(def);
                        if (defnode == null && !Decomposition.getConcepts2Ignore().contains(def)) {
                            defnode = addConceptRecursivly(def, threshold, nodeType); //TODO: have a specific threshold for definitions
                        }
                        if (defnode != null) {
                            DefinitionLink link = new DefinitionLink();
                            link.setSource(node);
                            link.setTarget(defnode);
                            link.setWeight(MarkerPassingConfig.getDefinitionLinkWeight());
                            if (!defnode.getLinks().contains(link)) {
                                node.addLink(link);
                            }
                        }
                    }
                }
            }
            for (Concept antonym : concept.getAntonyms()) {
                if (antonym != null) {
                    Node antoNode = nodes.get(antonym);
                    if (antoNode == null) {
                        antoNode = addConceptRecursivly(antonym, threshold, nodeType); //TODO: have a specific threshold for hypernyms
                    }
                    AntonymLink link = new AntonymLink();
                    link.setSource(node);
                    link.setTarget(antoNode);
                    link.setWeight(MarkerPassingConfig.getAntonymLinkWeight());
                    if (!node.getLinks().contains(link)) {
                        node.addLink(link);
                    }
                }
            }
            for (Concept arbitraryRelation : concept.getArbitraryRelations()) {
                if (arbitraryRelation != null) {
                    Node relatedNode = nodes.get(arbitraryRelation);
                    if (relatedNode == null) {
                        relatedNode = addConceptRecursivly(arbitraryRelation, threshold, nodeType); //TODO: have a specific threshold for arbitrary relationships
                    }
                    ArbitraryRelationLink link = new ArbitraryRelationLink();
                    link.setSource(node);
                    link.setTarget(relatedNode);
                    link.setWeight(MarkerPassingConfig.getDefaultArbitraryRelationLinkWeight());
                    link.setRelationName(arbitraryRelation.getOriginatedRelationName());
                    if (!node.getLinks().contains(link)) {
                        node.addLink(link);
                    }
                }
            }

            return node;
        }
    }

    /**
     * Example implementation of the Node interface for nodes which have a double threshold.
     */


    public void addMarkerToNode(Concept concept2Activate, Marker activationMarker) {
        Node node2activate = nodes.get(concept2Activate);
        node2activate.getMarkers().add(activationMarker);
        getActiveNodes().add(node2activate);
    }


}

File: semantic_decomposition/graph/spreading_activation/marker_passing/double_marker_passing.py
from __future__ import annotations
from typing import Collection, Dict, List, Optional, Set, TYPE_CHECKING

from ....marker_passing.spreading_algorithm import SpreadingAlgorithm
from ....marker_passing.spreading_step import SpreadingStep
from ....marker_passing.in_function import InFunction
from ....marker_passing.out_function import OutFunction
from ....marker_passing.select_firing_nodes_function import SelectFiringNodesFunction
from ....marker_passing.node import Node
from ...semantic_net import SemanticNet
from ..count_termination_condition import CountTerminationCondition

if TYPE_CHECKING:
    from ....concept import Concept
    from ....entities.spreading_activation.marker_passing_config import MarkerPassingConfig


class _FireThresholdNodes(SelectFiringNodesFunction):
    def __init__(self, config: Optional["MarkerPassingConfig"]) -> None:
        self._config = config

    def compute(self, active_nodes: Collection[Node]) -> Collection[Node]:
        from ....entities.nodes.double_node_with_multiple_thresholds import (
            DoubleNodeWithMultipleThresholds,
        )
        result = []
        for node in active_nodes:
            if isinstance(node, DoubleNodeWithMultipleThresholds):
                if node.check_thresholds():
                    result.append(node)
        return result


class _DoubleMultiIn(InFunction):
    def compute(self, input_steps: Collection[SpreadingStep], node: Node) -> None:
        if hasattr(node, "in_function"):
            node.in_function(input_steps)


class _DoubleMultiOut(OutFunction):
    def compute(self, node: Node) -> Collection[SpreadingStep]:
        if hasattr(node, "out_function"):
            return node.out_function()
        return []


class DoubleMarkerPassing(SpreadingAlgorithm):
    """
    Marker-passing algorithm using DoubleNodeWithMultipleThresholds nodes.

    Replaces the JGraphT-backed Java implementation; uses SemanticNet instead.
    """

    def __init__(
        self,
        config: Optional["MarkerPassingConfig"] = None,
    ) -> None:
        super().__init__()
        from ....entities.spreading_activation.marker_passing_config import MarkerPassingConfig
        self._config = config or MarkerPassingConfig()
        self._graph: SemanticNet = SemanticNet()
        # Bidirectional map: Concept <-> Node
        self._concept_to_node: Dict["Concept", Node] = {}
        self._node_to_concept: Dict[Node, "Concept"] = {}

        termination = CountTerminationCondition(self._config.termination_pulse_count)
        self.termination_condition = termination
        self.preprocessing_steps = [termination]
        self.in_function = _DoubleMultiIn()
        self.out_function = _DoubleMultiOut()
        self.select_firing_nodes = _FireThresholdNodes(self._config)

    def fill_nodes(self, concepts: List["Concept"]) -> None:
        for concept in concepts:
            self._add_concept_recursively(concept, self._config.decomposition_depth)
        self._active_nodes = list(self._concept_to_node.values())

    def _add_concept_recursively(self, concept: "Concept", depth: int) -> Node:
        if concept in self._concept_to_node:
            return self._concept_to_node[concept]

        from ....entities.nodes.double_node_with_multiple_thresholds import (
            DoubleNodeWithMultipleThresholds,
        )
        node = DoubleNodeWithMultipleThresholds(
            litheral=concept.litheral, concept=concept, config=self._config
        )
        self._concept_to_node[concept] = node
        self._node_to_concept[node] = concept
        self._graph.add_vertex(node)

        if depth <= 0:
            return node

        self._connect_related(concept, concept.synonyms, "SynonymLink", depth)
        self._connect_related(concept, concept.antonyms, "AntonymLink", depth)
        self._connect_related(concept, concept.hypernyms, "HypernymLink", depth)
        self._connect_related(concept, concept.hyponyms, "HyponymLink", depth)
        self._connect_related(concept, concept.meronyms, "WeightedLink", depth)
        for definition in concept.definitions:
            if hasattr(definition, "concepts"):
                self._connect_related(concept, definition.concepts, "DefinitionLink", depth)
        return node

    def _connect_related(
        self,
        source_concept: "Concept",
        related: List["Concept"],
        link_class: str,
        depth: int,
    ) -> None:
        from ....entities.links import (
            SynonymLink, AntonymLink, HypernymLink, HyponymLink, DefinitionLink, WeightedLink,
        )
        link_map = {
            "SynonymLink": SynonymLink,
            "AntonymLink": AntonymLink,
            "HypernymLink": HypernymLink,
            "HyponymLink": HyponymLink,
            "DefinitionLink": DefinitionLink,
            "WeightedLink": WeightedLink,
        }
        LinkClass = link_map.get(link_class, WeightedLink)
        source_node = self._concept_to_node[source_concept]
        for target_concept in related:
            target_node = self._add_concept_recursively(target_concept, depth - 1)
            link = LinkClass(source=source_node, target=target_node, config=self._config)
            if link not in source_node.get_links():
                source_node.get_links().append(link)

    @staticmethod
    def do_initial_marking(
        nodes: Dict["Concept", Node],
        concepts: List["Concept"],
        start_activation: float,
    ) -> None:
        from ....entities.markers.double_marker_with_origin import DoubleMarkerWithOrigin
        for concept in concepts:
            node = nodes.get(concept)
            if node is not None:
                marker = DoubleMarkerWithOrigin(
                    activation=start_activation, origin=concept
                )
                node.add_marker(marker)
                from ....entities.nodes.double_node_with_multiple_thresholds import (
                    DoubleNodeWithMultipleThresholds,
                )
                if isinstance(node, DoubleNodeWithMultipleThresholds):
                    node.activation[concept][start_activation].append(concept)
